presto-main/src/main/java/com/facebook/presto/operator/LookupJoiner.java

```diff
@@ -0,0 +1,247 @@
+        if (lookupSource != null) {
+            lookupSource.close();
+        }
+    }
+
+    /**
+     * Produce rows matching join condition for the current probe position. If this method was called previously
+     * for the current probe position, calling this again will produce rows that wasn't been produced in previous
+     * invocations.
+     *
+     * @return true if all eligible rows have been produced; false otherwise (because pageBuilder became full)
+     */
+    private boolean joinCurrentPosition(Counter lookupPositionsConsidered)
+    {
+        // while we have a position on lookup side to join against...
+        while (joinPosition >= 0) {
+            lookupPositionsConsidered.increment();
+            if (lookupSource.isJoinPositionEligible(joinPosition, probe.getPosition(), probe.getPage())) {
+                currentProbePositionProducedRow = true;
+
+                pageBuilder.declarePosition();
+                // write probe columns
+                probe.appendTo(pageBuilder);
+                // write build columns
+                lookupSource.appendTo(joinPosition, pageBuilder, probe.getOutputChannelCount());
+            }
+
+            // get next position on lookup side for this probe row
+            joinPosition = lookupSource.getNextJoinPosition(joinPosition, probe.getPosition(), probe.getPage());
+
+            if (lookupPositionsConsidered.get() >= MAX_POSITIONS_EVALUATED_PER_CALL) {
+                return false;
+            }
+            if (pageBuilder.isFull()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @return whether there are more positions on probe side
+     */
+    private boolean advanceProbePosition()
+    {
+        if (!probe.advanceNextPosition()) {
+            probe = null;
+            return false;
+        }
+
+        // update join position
+        joinPosition = probe.getCurrentJoinPosition();
+        return true;
+    }
+
+    /**
+     * Produce a row for the current probe position, if it doesn't match any row on lookup side and this is an outer join.
+     *
+     * @return whether pageBuilder became full
+     */
+    private boolean outerJoinCurrentPosition()
+    {
+        if (probeOnOuterSide && joinPosition < 0) {
+            // write probe columns
+            pageBuilder.declarePosition();
+            probe.appendTo(pageBuilder);
+
+            // write nulls into build columns
+            int outputIndex = probe.getOutputChannelCount();
+            for (int buildChannel = 0; buildChannel < lookupSource.getChannelCount(); buildChannel++) {
+                pageBuilder.getBlockBuilder(outputIndex).appendNull();
+                outputIndex++;
+            }
+            if (pageBuilder.isFull()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    // This class needs to be public because LookupJoinOperator is isolated.
```

Comment:
I'm not sure. I have done so, but maybe @dain could you take a deeper look into this aspect?